report/src/main/java/com/hartwig/actin/report/pdf/tables/molecular/RecentMolecularSummaryGenerator.java

```diff
@@ -36,113 +37,162 @@ public class RecentMolecularSummaryGenerator implements TableGenerator {
     @NotNull
-    private  Set<String> eventsForEligibleTrials() {
-        Set<String> molecularEvents = Sets.newTreeSet(Ordering.natural());
-        for (EvaluatedTrial trial : trials) {
-            if (trial.isPotentiallyEligible() && trial.isOpen()) {
-                molecularEvents.addAll(trial.molecularEvents());
+    private Cell biopsySummary() {
+        String biopsyLocation = clinical.tumor().biopsyLocation();
+        Double purity = molecular.characteristics().purity();
+        if (biopsyLocation != null) {
+            if (purity != null) {
+                Text biopsyText = new Text(biopsyLocation).addStyle(Styles.tableHighlightStyle());
+                Text purityText = new Text(String.format(ApplicationConfig.LOCALE, " (purity %d%%)", Math.round(purity * 100)));
+                purityText.addStyle(molecular.hasSufficientQuality() ? Styles.tableNoticeStyle() : Styles.tableHighlightStyle());
+                return Cells.create(new Paragraph().addAll(Arrays.asList(biopsyText, purityText)));
+            } else {
+                return Cells.createValue(biopsyLocation);
             }
+        } else {
+            return Cells.createValue(Formats.VALUE_UNKNOWN);
         }
-        return molecularEvents;
     }
 
     @NotNull
-    private static String biopsyLocation(@NotNull TumorDetails tumor) {
-        String biopsyLocation = tumor.biopsyLocation();
-        return biopsyLocation != null ? biopsyLocation : Formats.VALUE_UNKNOWN;
+    private <T extends GeneAlteration & Driver> Optional<String> summaryStringOptionForGeneAlterations(Stream<T> geneAlterationStream) {
+        String genes = geneAlterationStream.filter(variant -> variant.driverLikelihood() == DriverLikelihood.HIGH)
+                .map(GeneAlteration::gene)
+                .collect(Collectors.joining(", "));
+        return Optional.of(genes.isEmpty() ? Formats.VALUE_NONE : genes);
     }
 
     @NotNull
-    private static Cell createPredictedTumorOriginValue(@Nullable PredictedTumorOrigin predictedTumorOrigin) {
-        String interpretation = TumorOriginInterpreter.interpret(predictedTumorOrigin);
-        if (TumorOriginInterpreter.hasConfidentPrediction(predictedTumorOrigin)) {
-            return Cells.createValue(interpretation);
-        } else {
-            return Cells.createValueWarn("Inconclusive, highest prediction: " + interpretation);
-        }
+    private Optional<String> genesWithHighDriverMutationStringOption() {
+        return summaryStringOptionForGeneAlterations(molecular.drivers().variants().stream());
     }
 
     @NotNull
-    private static Cell addIndent(@NotNull Cell cell) {
-        return cell.setPaddingLeft(10);
+    private Optional<String> genesWithHighDriverAmplificationStringOption() {
+        return summaryStringOptionForGeneAlterations(molecular.drivers()
+                .copyNumbers()
+                .stream()
+                .filter(copyNumber -> copyNumber.type().isGain() && !copyNumber.type().isLoss()));
```

Comment:
It's a weakness in the datamodel that you need to check it this way. A gain can never be a loss, but that's not enforced. Maybe "partial" should be a boolean in the main copy number datamodel, such type can be either "GAIN", "LOSS", "NONE". Or for now just make a TODO note / jira ticket and leave for later? Best not combine this ticket with datamodel changes I imagine...

Reply:
I'll go ahead and do this.

Reply:
If I recall correctly, I added isGain() to simplify the code that you propose now, but maybe should revert that...?

Reply:
`CopyNumber`s are always created from a type (`FULL_GAIN, PARTIAL_GAIN, LOSS, NONE`), and this operation can never result in a `CopyNumber` with both `isGain` and `isLoss` set to true. In that case, it should be sufficient to just check `isGain()` for amplifications and `isLoss()` for deletions. Do you agree?

Reply:
In theory someone could add one more enum value that sets both to true, but that would be a pretty huge mistake by itself so maybe fine to assume that won't happen. I am fine either way, if you filter amps just by "isGain == true" then I am quite confident that will never be a problem